### ISSUE TITLE
Requests made when a JWT token expires always fail (Supabase v2) 

### ISSUE THREAD

**Issue by alexreyes** [supabase/gotrue-js#487]
# Bug report

## Describe the bug

Making a request right when a JWT token expires results in a `400: JWT expired` error on React Native iOS. This only happens exactly as the token is expiring, and subsequent requests seem to work. 

I think this might also be causing the random logging out issues as well. 

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Set the JWT expiry timeout to 5 seconds
2. Make a request as the JWT token expires (every 5 seconds) 
3. See JWT expired error message

NOTE: making a second request after getting the JWT expired error later results in that second request working. 

## Expected behavior

I would expect the token to be refreshed before making the request and the request to work fine.

## Screenshots

<img width="737" alt="Screen Shot 2022-10-13 at 11 10 41 AM" src="https://user-images.githubusercontent.com/7016669/195673726-255b9f21-647b-4d62-ad80-39f0ed7725d2.png">

## System information

- OS: macOS
- Browser: React native iOS
- Version of supabase-js: ^2.0.0
- Version of Node.js: v16.13.1

**Comment by neocho** [comment]
Good question: curious about this too

**Comment by hf** [comment]
Hey we're tracking this issue internally. We did some changes to the way expiration is handled in v2 (i.e. removed a safety margin that was previously there). But I think it's time to bring it back now, as it clearly is not behaving as well as we hoped. I'll submit a PR right now.

**Comment by alexreyes** [comment]
@hf okay thanks! When you submit the PR, could you please link it here? Would love to test it and give feedback (and maybe switch to the PR branch if it fixes it). 

For context, this is causing data loss for our users where they make an action on the app and don't see it reflected once they refresh/reopen/etc

**Comment by hf** [comment]
This is the PR (that was contributed by another member of the community a few days ago). https://github.com/supabase/gotrue-js/pull/482

**Comment by hf** [comment]
If you could leave your review on it that would be great since it's going to take at least one day before we can deploy a fix to this.

**Comment by hf** [comment]
Hey @alexreyes, just sanity checking -- you're setting the JWT expiry to 5 seconds only to test this behavior right? We don't recommend setting an expiry time of less than 2-3 minutes on it, except for testing purposes like this.

**Comment by alexreyes** [comment]
Yup, it’s just for testing. Normally I set the expiry to 1 hour

**Comment by alexreyes** [comment]
@hf I tried it, and I still had the same problem of the request failing because of JWT expiry when using the changes linked in the PR. Do you have any other ideas what might be causing this?

**Comment by alexreyes** [comment]
I ended up figuring out why this bug was happening! Turns out, it had nothing to do with Supabase and instead was due to the way react renders things. 

The situation: 

I am updating the JWT token in a useEffect() in the context file for my project. There are times when the JWT token updates in the context, and this updated token isn't being passed down and reflected in the child component where the API request is made. 

The solution:

 I use ```await supabase.auth.getSession();``` to retrieve the JWT token before making API calls instead of getting the session from the context
